File: src/main/java/net/moltendorf/Bukkit/IntelliDoors/Door.java
package net.moltendorf.Bukkit.IntelliDoors;

import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;

/**
 * Created by moltendorf on 15/05/23.
 *
 * @author moltendorf
 */
public class Door extends DoorType {
	final protected Block top, bottom;

	protected byte topData, bottomData;

	public Door(final Block block) {
		// org.bukkit.material.Door: Deprecated? Is this really so hard? Sigh.
		final byte data = block.getData();

		if ((data & 8) == 8) {
			// Top of door.
			top = block;
			bottom = block.getRelative(BlockFace.DOWN);

			topData = data;
			bottomData = bottom.getData();
		} else {
			// Bottom of door.
			top = block.getRelative(BlockFace.UP);
			bottom = block;

			topData = top.getData();
			bottomData = data;
		}
	}

	private static BlockFace[] Facing = {BlockFace.SOUTH, BlockFace.WEST, BlockFace.NORTH, BlockFace.EAST};

	public BlockFace getFacing() {
		return Facing[bottomData & 3];
	}

	public boolean isLeft() {
		return (topData & 1) == 0;
	}

	public boolean isRight() {
		return (topData & 1) == 1;
	}

	public boolean isClosed() {
		return (bottomData & 4) == 0;
	}

	public boolean isOpened() {
		return (bottomData & 4) == 4;
	}

	public void close() {
		if (isOpened()) {
			bottomData -= 4;

			bottom.setData(bottomData);
		}
	}

	public void open() {
		if (isClosed()) {
			bottomData += 4;

			bottom.setData(bottomData);
		}
	}
}

File: src/main/kotlin/net/moltendorf/bukkit/intellidoors/controller/Door.kt
package net.moltendorf.bukkit.intellidoors.controller

import net.moltendorf.bukkit.intellidoors.*
import net.moltendorf.bukkit.intellidoors.controller.Door.Flag.*
import net.moltendorf.bukkit.intellidoors.settings.*
import org.bukkit.block.*
import org.bukkit.block.BlockFace.*

/**
 * Created by moltendorf on 2016-06-03.
 *
 * Boiler plate class that provides base logic usable by all doors.
 */
abstract class Door(override val settings : Settings) : DoorInterface {
  override fun equals(other : Any?) : Bool {
    if (this === other) {
      return true
    }

    if (other?.javaClass != javaClass) {
      return false
    }

    other as DoorInterface

    if (location != other.location) {
      return false
    }

    return true
  }

  override fun hashCode() = location.hashCode()

  companion object {
    val faces = listOf(NORTH, EAST, SOUTH, WEST, NORTH, EAST, SOUTH)
    val lookup = mapOf(Pair(NORTH, 0), Pair(EAST, 1), Pair(SOUTH, 2), Pair(WEST, 3))

    fun rotate(value : BlockFace, amount : Int) = faces[lookup[value]!! + amount]
  }

  enum class Flag(val value : Int) : (BlockState) -> Unit {
    OPEN(4),
    TOP(8);

    override fun invoke(p1 : BlockState) {
      p1.intData = p1.intData or value
    }
  }

  enum class Mask(flag : Flag) : (BlockState) -> Unit {
    CLOSED(OPEN),
    BOTTOM(TOP);

    val value = flag.value.inv()

    override fun invoke(p1 : BlockState) {
      p1.intData = p1.intData and value
    }
  }
}
